### PR TITLE
Harden MiscellaneousFilesWorkspace against events from other threads

### PR DESCRIPTION
**Pull request by jasonmalinowski**
We had a bunch of crashes against Update 3 where
MiscellaneousFilesWorkspace would go sideways when we were responding
to workspace changed events for buffers on a different thread. Commit
dfcfbaa6f443236b29e912b9a4e9f16045952c04 mitigated the direct crashes
somewhat, but still meant we had some racy code which we probably
didn't want. This hardens the code so if we do get a notification
on a background thread (which is unfortunately controlled by the
sending workspace), we'll just send it back to the UI thread and safely
handle it from there.

In the dumps in question, even the Roslyn workspace was attached to
the background thread, which is known issue if somebody created it via
MEF when they shouldn't have. That's being fixed as a part of
https://github.com/dotnet/roslyn/pull/14354 so we won't concern
ourselves with it here.

_Review:_ @dotnet/roslyn-ide. This is targeting **master** instead of dev15-rc because I'm reasonably convinced that dfcfbaa6f443236b29e912b9a4e9f16045952c04 is sufficient for that release. This is mostly just hardening since I've already gone through the mental gymnastics to reunderstand this code.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
restest windows_debug_unit64_prtest please

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs [hunk 1]
```diff
@@ -210,14 +220,31 @@ private void TrackOpenedDocument(uint docCookie, string moniker)
 
         private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
+            // We may or may not be getting this notification from the foreground thread if another workspace
+            // is raising events on a background. Let's send it back to the UI thread since.
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs]: since what?  Don't leave me in suspense...

**Comment by jasonmalinowski** [comment]
retest windows_debug_unit32_prtest please

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest this please

This hit #13243 which is now fixed.

**Comment by MRatnam** [comment]
Hi is a hotfix or patch available for VS2015?

**Comment by jasonmalinowski** [comment]
Yes, there is a patch available for this particular issue: https://aka.ms/kb3217593-download.

**Comment by MRatnam** [comment]
@jasonmalinowski Great thanks very much

**Pull request merged**